File: UCF HSPT Documents/2013/Solutions/slow.c

// 27th Annual UCF High School Programming Tournament
//
// Problem Name: You're Too Slow! (slow)
//
// Author:        Aaron Teetor
// Judge Data:    Antony Stabile
// C Solution:    Glenn Martin
// Java Solution: Danny Wasserman
// Verifier:      Patrick Fenelon

#include <stdio.h>
#include <math.h>


#define DEBUG   0

#define TRUE    1
#define FALSE   0


typedef struct
{
   int   x;
   int   y;
   int   enemies;
} Room;


int episodeLength(int episodeNum, int totalEpisodes)
{
   // Season finale has 30 minutes; others have 22
   if (episodeNum == totalEpisodes)
      return 30;
   else
      return 22;
}


double distance(Room r1, Room r2)
{
   double   d1;
   double   d2;

   // Compute the distance between the rooms and return it
   d1 = r1.x - r2.x;
   d2 = r1.y - r2.y;
   return sqrt(d1*d1 + d2*d2);
}


int main()
{
   FILE *   infile;
   int      numSims;
   int      i;
   int      numEpisodes;
   int      numRooms;
   int      j;
   Room     rooms[1000];
   double   timeInEpisode;
   int      location;
   int      episode;
   int      ok;
   int      episodeLen;
   double   travelTime;
   int      numEnemies;
   double   dist;

   // Open the input file for reading
   infile = fopen("slow.in", "r");

   // Read in the number of simulations to process
   fscanf(infile, "%d", &numSims);

   // Loop over each simulation
   for (i=0; i < numSims; i++)
   {
      // Read in the number of episodes left and number of rooms that exist
      fscanf(infile, "%d %d", &numEpisodes, &numRooms);

      // Read in the room data
      for (j=0; j < numRooms; j++)
      {
         fscanf(infile, "%d %d %d", &rooms[j].x, &rooms[j].y, 
                &rooms[j].enemies);
      }

      // Initialing starting state (time 0 into episode, first room, 
      // first episode)
      timeInEpisode = 0;
      location = 0;
      episode = 1;

      // Assume we're okay until we find that we aren't or we finished
      ok = TRUE;
      while ( (ok == TRUE) && (location < numRooms) )
      {
         if (DEBUG)
            printf("AT ROOM %d\n", location+1);

         // Initialize how long the current episode is
         episodeLen = episodeLength(episode, numEpisodes);

         // First, defeat enemies in this room (if we have shows left)
         numEnemies = rooms[location].enemies;
         while ( (numEnemies > 0) && (episode <= numEpisodes) )
         {
            // Do we have time to defeat the enemy?
            if (episodeLen - timeInEpisode < 3)
            {
               // No, we do not so we have to wait for an episode where
               // we have time
               while ( (episodeLen - timeInEpisode < 3) && 
                       (episode <= numEpisodes) )
               {
                  // Go to start of next episode
                  timeInEpisode = 0;
                  episode++;
                  episodeLen = episodeLength(episode, numEpisodes);
               }
            }
               
            // Now, defeat the enemy!
            if (DEBUG)
               printf("Defeated enemy from %f to %f\n", timeInEpisode,
                      timeInEpisode + 3);
            timeInEpisode += 3;
            numEnemies--;
         }

         // We defeated everybody in the room so it's time to go to the
         // next room (if it exists)
         if (location < numRooms - 1)
         {
            // Get distance to next room
            dist = distance(rooms[location], rooms[location+1]);
            if (DEBUG)
               printf("Need to travel %f km\n", dist);

            // Determine how long it will take to get there (account for
            // acceleration time)
            if (dist > 10)
            {
               // Sonic will accelerate to his constant maximum speed
               travelTime = 1.0 + (dist - 10) / 20;
            }
            else
            {
               // Sonic arrives while he's still accelerating
               travelTime = dist / 10;
            }
            if (DEBUG)
               printf("It will take %f minutes\n", travelTime);

            // Now, that we now how long it will take to get to the next
            // room, use time in the episodes to do the travel
            while ( (episodeLen - timeInEpisode < travelTime) && 
                    (episode <= numEpisodes) )
            {
               // Use up what's left of this episode
               if (DEBUG)
                  printf("Traveling for %f minutes\n", 
                         (episodeLen - timeInEpisode));
               travelTime -= (episodeLen - timeInEpisode);
               timeInEpisode = 0;
               episode++;
               episodeLen = episodeLength(episode, numEpisodes);
            }

            // And finish leftover travel in this episode if there is any left
            if (travelTime > 0)
            {
               // Finish the travel in the time we need
               if (DEBUG)
                  printf("Traveling for %f minutes\n", travelTime);
               timeInEpisode += travelTime;
               travelTime = 0;
            }
         }

         // Either way, we need to advance to the next room
         location++;

         // Check to see if we ran out of episodes
         if (DEBUG)
             printf("Finished at %f minutes into episode %d\n", timeInEpisode,
                    episode);
         if (episode > numEpisodes)
         {
            // We did, uh oh!
            ok = FALSE;
         }
      }

      // Output accordingly
      if (ok == TRUE)
         printf("Simulation #%d: Abort fortress!\n", i+1);
      else
         printf("Simulation #%d: You're too slow!\n", i+1);
   }

   // Return that we're ending okay
   fclose(infile);
   return 0;
}

File: UCF HSPT Team/src/y2013/slow.java

// 27th Annual UCF High School Programming Tournament
//
// Problem Name: You're Too Slow! (slow)
//
// Author:        Aaron Teetor
// Judge Data:    Antony Stabile
// C Solution:    Glenn Martin
// Java Solution: Danny Wasserman
// Verifier:      Patrick Fenelon

import java.io.File;
import java.io.IOException;
import java.util.Scanner;


public class slow {
  public static void main(String[] args) throws IOException {
    // Open input file.
    Scanner in = new Scanner(new File("slow.in"));

    // Read in number of simulations and loop over them.
    int simulations = in.nextInt();
    for(int simOn = 1; simOn <= simulations; simOn++) {
      // Read number of episodes left in the season.
      int numEpisodes = in.nextInt();
      // Read number of rooms in this season.
      int rooms = in.nextInt();
      
      // Create space for rooms and enemies.
      Point[] roomLocations = new Point[rooms];
      int[] numEnemies = new int[rooms];

 
      // Read in each room.
      for(int i = 0; i < rooms; i++) {
        roomLocations[i] = new Point(in.nextInt(), in.nextInt());
        numEnemies[i] = in.nextInt();
      }
      
      // Preprocess the time it takes to get from 1 room to another.
      // Time is stored in minutes.
      double[] timeToMove = new double[rooms];
      for(int i = 0; i < rooms-1; i++) {
        // Find the distance between the two rooms.
        double distance = roomLocations[i].distance(roomLocations[i+1]);
        
        // Compute time it takes to get to the next room (handle sonic 
        // accelerating).
        double time = 0;
        if(distance <= 10) {
          // entire time sonic moves at 10 km/min.
          time = distance / 10;
        } else {
          time = 1;
          distance -= 10;
          // After the first 10 KM he moves at 20 km/min.
          time += distance / 20;
        }

        // Store the final time.
        timeToMove[i] = time;
      }
      
      // The time to move from the last room to a 'final' room is zero.
      // this makes it easy to detect if Sonic finished the labyrinths.
      
      // Start in first room.
      int roomOn = 0;

      // Process each episode.
      for(int epOn = 1; epOn <= numEpisodes; epOn++) {
        // Figure out length of this episode.
        double time = 22; 
        if(epOn == numEpisodes) {
          // Season finale has 30 minutes.
          time = 30; 
        }

        // Flag for seeing if something was processed during the 
        // next loop for the episode.
        boolean changed = true;

        // Loop while something changes
        while(changed) {
          // Default that nothing will change this loop
          changed=false;

          // While there are enemies that need to be defeated, defeat one!
          while(numEnemies[roomOn] > 0) {
            // If 3 minutes are not available, there is not enough time.
            if(time < 3) break;

            // Defeat the enemy!
            changed=true; // enemy defeated, sonic moved on!
            time-=3; // it takes 3 minutes to defeat an enemy.
            numEnemies[roomOn]--; // an enemy was defeated, update.
          }
          
          // if there are no more enemies, time to move on.
          if(numEnemies[roomOn]==0) {
            // For this part, we can consider ourselves to still be in 
            // the current room and there is just a timer running that 
            // is the time to the next room.
            if(timeToMove[roomOn] <= time) {
              changed=true; //sonic changed rooms!
              time-=timeToMove[roomOn]; // took time to get there.
              roomOn++; // move rooms!
            } else {
              // If he doesn't change rooms...
              timeToMove[roomOn]-=time; // use the timer for running.
              time=0;
            }
          }

          // If sonic finished, break out.
          if(roomOn==rooms) break;
        }

        // If sonic finished, break out.
        if(roomOn==rooms) break;
      }
      
      // If sonic made it through all the rooms (numbered from 0...(rooms-1)) 
      // then he should be in 'rooms' since it takes no time to move to 
      // the 'final' room.
      if(roomOn == rooms) {
        System.out.printf("Simulation #%d: Abort fortress!\n",simOn);
      } else {
        System.out.printf("Simulation #%d: You're too slow!\n",simOn);
      }
    }

    // Close scanner
    in.close();
  }
  

  // Point class used to represent where a room is.
  static class Point {
    long x, y;
    public Point(long x, long y) {
      this.x = x;
      this.y = y;
    }
    
    // return the distance between two points.
    // a^2 + b^2 = c^2 ==> c = Sqrt(a^2 + b^2).
    // a is the x distance and b is the y distance.
    public double distance(Point o) {
      return Math.sqrt( (x-o.x)*(x-o.x) + (y-o.y)*(y-o.y) );
    }
    
    public String toString() {
      return String.format("(%d, %d)",x,y);
    }
  }
}
